internal/uvm/scsi.go

```diff
@@ -26,67 +25,142 @@ var (
 // SCSI controllers associated with a utility VM to use.
 // Lock must be held when calling this function
-func (uvm *UtilityVM) allocateSCSI(ctx context.Context, hostPath string, uvmPath string, isLayer bool) (int, int32, error) {
+func (uvm *UtilityVM) allocateSCSI(ctx context.Context, hostPath string, uvmPath string, isLayer bool) (*SCSIMount, error) {
 	for controller, luns := range uvm.scsiLocations {
 		for lun, si := range luns {
-			if si.hostPath == "" {
-				uvm.scsiLocations[controller][lun].hostPath = hostPath
-				uvm.scsiLocations[controller][lun].uvmPath = uvmPath
-				uvm.scsiLocations[controller][lun].isLayer = isLayer
-				uvm.scsiLocations[controller][lun].refCount = 1
-				log.G(ctx).WithFields(logrus.Fields{
-					"hostPath":   hostPath,
-					"uvmPath":    uvmPath,
-					"isLayer":    isLayer,
-					"refCount":   1,
-					"controller": controller,
-					"lun":        lun,
-				}).Debug("allocated SCSI location")
-				return controller, int32(lun), nil
+			// If si is nil, we have found an open slot so we allocate a new SCSIMount
+			if si == nil {
+				uvm.scsiLocations[controller][lun] = &SCSIMount{
+					HostPath:   hostPath,
+					UvmPath:    uvmPath,
+					isLayer:    isLayer,
+					refCount:   1,
+					Controller: controller,
+					Lun:        int32(lun),
+				}
+				log.G(ctx).Debug("allocated SCSI location: ", uvm.scsiLocations[controller][lun])
+				return uvm.scsiLocations[controller][lun], nil
 			}
 		}
 	}
-	return -1, -1, ErrNoAvailableLocation
+	return nil, ErrNoAvailableLocation
 }
 
-func (uvm *UtilityVM) deallocateSCSI(ctx context.Context, controller int, lun int32) {
+func (uvm *UtilityVM) deallocateSCSI(ctx context.Context, sm *SCSIMount) {
 	uvm.m.Lock()
 	defer uvm.m.Unlock()
-	si := uvm.scsiLocations[controller][lun]
-	if si.hostPath != "" {
-		log.G(ctx).WithFields(logrus.Fields{
-			"hostPath":   si.hostPath,
-			"uvmPath":    si.uvmPath,
-			"isLayer":    si.isLayer,
-			"refCount":   si.refCount,
-			"controller": controller,
-			"lun":        lun,
-		}).Debug("removed SCSI location")
-		uvm.scsiLocations[controller][lun] = scsiInfo{}
+	if sm != nil {
+		log.G(ctx).Debug("removed SCSI location: ", sm)
+		sm = nil
 	}
 }
 
 // Lock must be held when calling this function.
-func (uvm *UtilityVM) findSCSIAttachment(ctx context.Context, findThisHostPath string) (int, int32, string, error) {
-	for controller, luns := range uvm.scsiLocations {
-		for lun, si := range luns {
-			if si.hostPath == findThisHostPath {
-				log.G(ctx).WithFields(logrus.Fields{
-					"hostPath":   si.hostPath,
-					"uvmPath":    si.uvmPath,
-					"isLayer":    si.isLayer,
-					"refCount":   si.refCount,
-					"controller": controller,
-					"lun":        lun,
-				}).Debug("found SCSI location")
-				return controller, int32(lun), si.uvmPath, nil
+func (uvm *UtilityVM) findSCSIAttachment(ctx context.Context, findThisHostPath string) (*SCSIMount, error) {
+	for _, luns := range uvm.scsiLocations {
+		for _, si := range luns {
```

Comment:
I would use something like `sm` as the enumeration variable now that the type isn't called `scsiInfo` anymore.